internal/databases/postgres/copy_tar_ball_composer.go

```diff
@@ -0,0 +1,249 @@
+			prevFileTar[fileName] = tarName
+			prevTarFileSets[tarName] = append(prevTarFileSets[tarName], fileName)
+		}
+		tarUnchangedFilesCount[tarName] = len(fileSet)
+	}
+	files := &RegularBundleFiles{}
+	tarBallFilePacker := newTarBallFilePacker(bundle.DeltaMap,
+		bundle.IncrementFromLsn, files, maker.filePackerOptions)
+	return NewCopyTarBallComposer(bundle.TarBallQueue, tarBallFilePacker, files,
+		bundle.Crypter, maker.previousBackup, maker.newBackupName, tarUnchangedFilesCount,
+		prevFileTar, prevTarFileSets), nil
+}
+
+func (c *CopyTarBallComposer) AddFile(info *ComposeFileInfo) {
+	var fileName = info.header.Name
+	var currFile = fileInfo{}
+	if !c.prevBackup.SentinelDto.Files[fileName].MTime.Equal(info.header.ModTime) {
+		c.tarUnchangedFilesCount[c.prevFileTar[fileName]] = -1
+		currFile.status = doNotCopy
+	} else if _, exists := c.prevFileTar[fileName]; exists {
+		c.tarUnchangedFilesCount[c.prevFileTar[fileName]] -= 1
+		currFile.status = possibleCopy
+	} else {
+		currFile.status = fromNew
+	}
+	currFile.info = info
+	c.fileInfo[fileName] = &currFile
+}
+
+func (c *CopyTarBallComposer) AddHeader(fileInfoHeader *tar.Header, info os.FileInfo) error {
+	var fileName = fileInfoHeader.Name
+	var currHeader = headerInfo{}
+	if !c.prevBackup.SentinelDto.Files[fileName].MTime.Equal(info.ModTime()) {
+		c.tarUnchangedFilesCount[c.prevFileTar[fileName]] = -1
+		currHeader.status = doNotCopy
+	} else if _, exists := c.prevFileTar[fileName]; exists {
+		c.tarUnchangedFilesCount[c.prevFileTar[fileName]] -= 1
+		currHeader.status = possibleCopy
+	} else {
+		currHeader.status = fromNew
+	}
+	currHeader.info = info
+	currHeader.fileInfoHeader = fileInfoHeader
+	c.files.AddFile(fileInfoHeader, info, false)
+	c.headerInfos[fileName] = &currHeader
+	return nil
+}
+
+func (c *CopyTarBallComposer) SkipFile(tarHeader *tar.Header, fileInfo os.FileInfo) {
+	c.files.AddSkippedFile(tarHeader, fileInfo)
+}
+
+func (c *CopyTarBallComposer) copyTar(tarName string) {
+	tracelog.InfoLogger.Printf("Copying %s ...\n", tarName)
+	c.prevBackup.Folder.CopyObject(
+		utility.BaseBackupPath,
+		c.prevBackup.Name+internal.TarPartitionFolderName+tarName,
+		c.newBackupName+internal.TarPartitionFolderName+tarName)
+	c.copiedTars[tarName] = true
+	c.tarBallQueue.TarBallMaker.AddCopiedTarName(tarName)
+	for _, fileName := range c.prevTarFileSets[tarName] {
+		if file, exists := c.fileInfo[fileName]; exists {
+			file.status = processed
+			c.tarFileSets[tarName] = append(c.tarFileSets[tarName], fileName)
+			c.files.AddFile(file.info.header, file.info.fileInfo, file.info.isIncremented)
+		} else if header, exists := c.headerInfos[fileName]; exists {
+			header.status = processed
+			c.tarFileSets[tarName] = append(c.tarFileSets[tarName], fileName)
+			c.files.AddFile(header.fileInfoHeader, header.info, false)
+		}
+	}
+}
+
+func (c *CopyTarBallComposer) getTarBall() internal.TarBall {
+	tarBall := c.tarBallQueue.Deque()
+	tarBall.SetUp(c.crypter)
```

Comment:
To get rid of collisions between old tarballs and new ones, you can create a new ones with the custom names, [like it is being done here](https://github.com/wal-g/wal-g/blob/196c92016f2ffb11b261f2e152f35ab066cdd6fb/internal/databases/postgres/bundle.go#L319).

Reply:
If we will move name resolver into copy composer, we cannot check tar name correctness [here](https://github.com/wal-g/wal-g/blob/5f14deb57b7b9679f4307839f205e73cc7167c88/internal/databases/postgres/backup_push_handler.go#L262)

Reply:
It doesn't seems to be a problem if we assign a prefix like `copy_****.tar.gz` to a copied tars.